elastic/kibana
# [Observability] Implements UI Metric telemetry (alongside existing telemetry)

Closes #40438

## Summary

**Important: Read #40438 to fully understand the motivation for doing this work.**

This uses a new `useTrackPageview` hook to track page view "visit" data across observability so that we are streamlined on the way we collect that data. There is also a `useTrackMetric` hook for tracking more generic metrics other than pageviews. We can build on this strategy to begin collecting other telemetry in a uniform way.

Note: This PR doesn't touch the existing telemetry collection in any of the apps. I think it would be best to leave that in place for at least one version and compare the data we get from the new metrics vs. the existing metrics. 

## How to test

Run this branch locally and visit the Kibana Advanced Settings to make sure you have telemetry collection ("Usage Data") enabled. 

<img width="600" alt="Screen Shot 2019-07-05 at 11 58 23 AM" src="https://user-images.githubusercontent.com/159370/60741535-d1991900-9f37-11e9-8ef1-fc6d7748435a.png">

Once you're sure it's enabled, click through the various pages of the 4 observability apps (Uptime, APM, Infra Metrics and Infra Logs). If you stay on some for longer than 15 seconds, refresh a few pages, etc., you'll end up with a good sampling of the different visit data collected.

After you've done this, go back to Kibana Advanced Settings and scroll back to the bottom where the Usage Data can be enabled/disabled. Click on the link that says "See an example of what we collect". That link generates a request to the `_stats` endpoint, using the same endpoint that the telemetry server will use to request data. The new data from this PR is collected inside `stack_stats.kibana.plugins.ui_metric`, within the keys `apm`, `uptime`, `infra_metrics`, and `infra_logs`.

<img width="1002" alt="Screen Shot 2019-07-05 at 3 11 25 PM" src="https://user-images.githubusercontent.com/159370/60741589-0d33e300-9f38-11e9-9b9a-e93da1ad6302.png">

## Sample data

Here's the data I generated from doing the above exercise:

```json
{
  "infra_metrics": [
    {
      "key": "visit__metrics_explorer__delayed_15000ms",
      "value": 1
    },
    {
      "key": "visit__metrics_explorer",
      "value": 1
    },
    {
      "key": "visit__inventory",
      "value": 1
    }
  ],
  "infra_logs": [
    {
      "key": "visit__stream__delayed_15000ms",
      "value": 1
    },
    {
      "key": "visit__stream",
      "value": 1
    }
  ],
  "apm": [
    {
      "key": "visit__transaction_overview__delayed_15000ms",
      "value": 1
    },
    {
      "key": "visit__error_group_overview__delayed_15000ms",
      "value": 1
    },
    {
      "key": "visit__error_group_details__delayed_15000ms",
      "value": 1
    },
    {
      "key": "visit__transaction_overview",
      "value": 3
    },
    {
      "key": "visit__error_group_overview",
      "value": 4
    },
    {
      "key": "visit__error_group_details",
      "value": 1
    },
    {
      "key": "visit__services_overview",
      "value": 5
    }
  ],
  "uptime": [
    {
      "key": "visit__monitor__delayed_15000ms",
      "value": 1
    },
    {
      "key": "visit__monitor",
      "value": 1
    },
    {
      "key": "visit__overview",
      "value": 1
    }
  ]
}
```
